### PR TITLE
[readyformerge] heroes

### PR DESCRIPTION
**Pull request by sevu**
The advantage of using separate unit types for heroes is that they have their custom wiki page, so you can see their portrait and modifications there. It's not making coding easier, rather the other way around.
I think we should keep consistence on giving the heroes separate unit types or not.

### CODE REVIEW

**Comment by konecnyjakub** [comment]
I will take better look at this after I return home (tommorow afternoon/evening).

**Comment by sevu** [comment]
Take your time.

**Review by konecnyjakub — COMMENTED** [review]
First and third commit looks fine. But [base_unit] also copies variations from the base unit which is not a good think imo.

**Comment by sevu** [comment]
That's true, but see below.
Basically I see 2 ways.
1) Going with [object]s for all heroes.
This would be the easier variant
2) Making one file with [base_unit], which contains all the heroes as variations, using different [variations] inside that to overwrite the original ones.
Would have the advantage of having a help entry.
The variation_name= field could be used for displaying the name of them. (though this doesn't work for merrin, but he could use another variation then)
And hide_help=yes for not displaying the unused variations, like we have no soulless goblin hero.

However, I think we should decide for one way and not mix it.

**Comment by sevu** [comment]
So that's what I had in mind. We have too less heroes to overwrite all variations, but [base_unit] has some features more than totally separate unit types.
Btw, variation_ids can be overridden too.

**Comment by konecnyjakub** [comment]
I was thinking about it and now I like the second idea more.

**Comment by konecnyjakub** [comment]
An idea: if there is special folder for heroes' unit types, it might be nice to move also existing unit types for heroes there (those that are not going to be deprecated). At least Wizard philosophy and Sophist philosophy, maybe also those Lord/Princess things.

**Comment by sevu** [comment]
Ok. I think I will leave this PR for a while until I get back to it, may be a while.

Moving the other heroes too sounds like a good idea.
As for the princess, in 1.13 there are new images, though still missing the attack animation. Going with the classical ones?

**Comment by konecnyjakub** [comment]
Alright.
For princess, I prefer the old sprites.

**Comment by sevu** [comment]
So, finally done, took a while...

One change, I gave Boreas only on level 1 +4HP compared to other bat variations, and later not. It was like that when Boreas used separate files, when switching to objects he had +4 HP on every level. There's not really a reason for this, feel free to adjust.

It would be now the time to use them in the scenario files. Could you do that?
S1 may serve as example. But the heroes will need an variation id which is different than the original ones.

**Comment by konecnyjakub** [comment]
> when switching to objects he had +4 HP on every level.

It seems that I have done it wrongly. My intention was to keep the amount of hitpoints on all levels same as before.

There is no problem with different variation ids because that change is not in a released version, it is just on GitHub.

But why are there 2 variations for Pech (drake and gryphon)? She should be based on gryphon and obviously I have made another mistake when I based her on drake variation.

Also, could you try to clean up the history? 3rd and 4th commit are not needed now, 5th and 7th commits can be squashed, instead of 2nd (merge) commit you can rebase the branch.
Actually, old unit types for Merrin and Neckfrill can be moved to 1_compatibility.cfg.

File: units/heroes/Running_Heroes.cfg
Comment (konecnyjakub) [inline units/heroes/Running_Heroes.cfg]: c) can be solved with this code:
> 
    [advancefrom]
        unit=Soulless Hero
        experience=35
    [/advancefrom]

**Comment by sevu** [comment]
So ... 2 of the 4 were easy doing, somehow I got rid of the merge commit too.
As for the squash - there is one commit in between, so it can't be squashed directly. I wanted to ask if you can cherry pick it, and i drop  it and squash the others. But .. now there's a merge conflict with that commit, thought the listed conflicts doesn't look conflicting...
When rebasing I had some earlier commits in the list too. I didn't change anything there, but I guess their commit IDs changed.
What are our options on merging this? Cherry-picking (however this is done) the one commit, resolving the conflict while doing that (should be no problem), and look if further conflicts arise?
Or can I remove the first two files from this PR? That would mean I have to get the old commit IDs... Sounds like being more problematic.
Well, I'm sure there are more problems ahead. I guess we'll see.

File: units/heroes/Running_Heroes.cfg
Comment (sevu) [reply units/heroes/Running_Heroes.cfg]: so this didn't get inherited... I added an experience key in the lvl1 file now

Comment (sevu) [reply units/heroes/Running_Heroes.cfg]: As for Boreas, I thought it was intended that you changed his movetype, since he actually looks like a drake. I added both variants, since there are anymore more variants than we need. Both would fit.

**Comment by sevu** [comment]
Hey, I think I finally came to the point where git starts making fun and stops being annoying :)

File: units/heroes/Running_Heroes.cfg
Comment (konecnyjakub) [reply units/heroes/Running_Heroes.cfg]: Well, we inherit from Running Corpse here which has this:

> [advancefrom]
        unit=Soulless
        experience=35
[/advancefrom]

You mean Pech, right :) ? Drake and gryphon variations do have same graphics but there is a difference in movement costs and defense between them. The gryphon variation is better for her and she was like that with her own unit types.

**Comment by konecnyjakub** [comment]
Good job with git here :+1: . Now I am going to check this in game.

**Comment by konecnyjakub** [comment]
Hm, it does not look so good. Merrin is fine but all other heroes have some issues as Soulless. Pech, Neckfrill and Spelk cannot advance to Running Corpse, Boreas is also just usual bat zombie (he too cannot advance). And all of them are actually shown as the base unit in help. However on higher levels they are good. Oh, and without this:
> [advancefrom]
    unit=Soulless Hero
    experience=35
[/advancefrom]

Running Hero is another advancement for Soulless but you already know that.

**Comment by konecnyjakub** [comment]
Imho, git is less annoying when you use a GUI. I can get things done with it more quickly but of course I use git from terminal on occasions. Some tips for git gui(I use/used for some time): TortoiseGit (it is actually just for Windows according to my knowledge), Git Cola (I use this now on Ubuntu).

**Comment by sevu** [comment]
I will take a look at it. Can you use the more difficult commands like git rebase, git rebase -i or cherry-pick with the GUI too? This would be super useful.

File: units/heroes/Running_Heroes.cfg
Comment (sevu) [reply units/heroes/Running_Heroes.cfg]: Yses, I mean Pech.  She could use gender=female too.
So, the advancefrom is inherited for all inherited variations, but not for the new one. I see.

**Comment by konecnyjakub** [comment]
I still cannot get the interactive rebase in Git Cola but normal rebase and cherry picking are fairly easy with that. I think that is also the case with TortoiseGit (it have been some time since I used it last time).

**Comment by konecnyjakub** [comment]
I discovered that adding advances_to=Running Hero to Soulless Hero's variations makes all the heroes but Boreas advance as intended (it is not surprising as he is not using the correct variation). But I still have no idea why they are shown as the base unit in help.
Adding gender and usage does not hurt, I guess, though the latter might be useless. But I do not get why you added so many empty [attack_anim] tags. They do not have so many attacks, do they?
The drake variation can be hid, it is not needed.

**Comment by sevu** [comment]
There is already an advances_to=Running Hero in the Soulless Hero file, which should be inherited to the others. As far as I got it the [advancefrom] in Running Non-Heroes file got inherited to the Running Heroes, weightling higher than the advances_to in the Soulless Hero file. Since the pastor variation is a new one it didn't get that [advancefrom] inherited. So far the story, like that is was earlier.

To fix advancement I added the [advancefrom] in the Running Heroes file, to overwrite the inherited [advancefrom]. At least that's what I thought it does... What you discovered was before I pushed that? Or with that change already?
Assuming that it was before:

What exactly did you discover, that adding the advance_to line explicitly in the [variations] in the Soulless Hero file, besides that it should get that already inherited from somewhere in the first lines of that file?
So, that would mean that the [advancefrom] in the Running Non-Hero file would actually overwrite sth. in the Soulless Non-Hero file rather than belonging in the Non-Hero file? But is for no reason not getting overwritten by the advance_to in the first lines of Soulless Hero file, but by an additional advance_to line in the variations?
That would mean... the [advancefrom] is sth. in the file in which it is not coded, and is overriding the advance_to there. (Did I so far get it right?)

So.... basically... [base_unit], [variation] and [advancefrom] are all the same. And we have three layers of overwriting.
And, used in the correct order, it can be a cycle of endless overwritings :dagger:
That's interesting.

Part of the fun of writing this code was to go to the limits of [variation] and [base_unit] by combining them, and looking if there is sth. which can be improved about them in mainline. These keys and the way the help is dealing with them seemed always a bit hacky to me.
Now I wonder what specifically could be improved in mainline to make things less hacky.

**Comment by sevu** [comment]
Adding gender and usage key is just for fun, hide_help will be inherited to all variations once we un-comment the part in the beginning of the file. The [attack_anim]s fix #14 , though I added one to much. It need to be 3 because the attack animation is realized using multiple [attack_anim]s (One for s, one for n, and one for se,sw,ne,ne).

**Comment by konecnyjakub** [comment]
I added [advancefrom] to Running Hero and advances_to to Soulless Hero's variation and it fixed the advancing problem for all heroes but Boreas. I had all your changes at that time.

> That would mean... the [advancefrom] is sth. in the file in which it is not coded, and is overriding the advance_to there. (Did I so far get it right?)

Seems so.

> It need to be 3 because the attack animation is realized using multiple [attack_anim]s (One for s, one for n, and one for se,sw,ne,ne).

Oh, right, I did not realise this.

**Comment by sevu** [comment]
>>It need to be 3 because the attack animation is realized using multiple [attack_anim]s (One for s, one for n, and one for se,sw,ne,ne).  

>Oh, right, I did not realise this.

me neither

Could you test if it works without the [advancefrom] in Running Heroes the same?  

What is different for Boreas, he's not using the correct variation? He *should* be not different than the others...

**Comment by konecnyjakub** [comment]
> Could you test if it works without the [advancefrom] in Running Heroes the same?

You mean if they advance to Running Hero when they just have advance_to in the variation? Yes, they do.

> He should be not different than the others...

Sure but he is. This all is strange.

**Comment by sevu** [comment]
> You mean if they advance to Running Hero when they just have advance_to in the variation? Yes, they do.
Ok, so things are like in the long post above.

about advancing boreas
-> try using a custom variation name for him, and on all levels the same
-> try using 'bat' as variation name

I had a problem placing hero variations using [unit] if the had the same variation name like the base_unit.
That looks like a bug in the engine.
Placing a hero in debug mode, changing variation with the :unit command and letting it advance works for me.

**Comment by konecnyjakub** [comment]
Finally got back to this. Sorry for the delay.

> try using a custom variation name for him, and on all levels the same

You mean variation id, right? (The variations' names are already different for all heroes.) I tried that but it was a complete disaster. As soulless, not only he could not advance, he actually had no sprite at all. But once I manually advanced him to running hero, he was just fine.

> try using 'bat' as variation name

It is this way now or am I missing anything?

**Comment by konecnyjakub** [comment]
Lately, I have been thinking a bit about Boreas' issues. Maybe we can workaround them by directly copying WML of the base unit for Soulless hero instead of using [base_unit]. I personally do not like this solution (because of the duplication of code) but as long as it works, I would go that way. Well, if it is not fixed on Wesnoth's end.

**Comment by sevu** [comment]
This we could.
For mainline I made a [feature request](https://gna.org/bugs/index.php?25624), thought the long text probably scared all devs away.

**Comment by konecnyjakub** [comment]
Are you going to do that soon or should I do that myself? And are you planning more changes here? I ask because I would like to release new version as soon as possible (and I am already 2 months behind eta :P).

**Comment by sevu** [comment]
I'll do it within the next days.
Since the code up until here is pretty well tested, I'll do no more rebasing, so in case this is going to be implemented in mainline we could revert the following commits.

**Comment by konecnyjakub** [comment]
Sounds good.

**Comment by sevu** [comment]
I previously missed you point about boreas, I just realized it when I applied that patch.

For the problem with wiki - what we could do would be the first approach (the one from before the rebase).
Though it may take a while to code it.

**Comment by konecnyjakub** [comment]
Does that __remove thing work on Wesnoth 1.12?
And with the first approach you mean this:

> Going with [object]s for all heroes.

?

**Comment by sevu** [comment]
I didn't test it for 1.12. For case it doesn't there would be one or two other possibilities.

I meant using separate unit types for all heroes, with having [base_unit] in them. And hiding the variations in them with hide_help=yes.

**Comment by konecnyjakub** [comment]
> I meant using separate unit types for all heroes, with having [base_unit] in them

That sounds quite messy to me. I would really prefer this: https://github.com/konecnyjakub/Danse_Macabre/pull/12#issuecomment-298169757. I did not try it yet but I am going to soon.

**Comment by sevu** [comment]
Boreas issue is solved for 1.12 too.

File: units/heroes/Brainless_Heroes.cfg
Comment (sevu) [inline units/heroes/Brainless_Heroes.cfg]: If we add hide_help=yes here, then the help is not working as expected anymore.
What do you think about how it is right now?

It can be worked around with having the heroes in separate files, which use [base_unit] on the main unit, and hide all inherited variations in them. The help system will have for these heroes then a link to the base unit, but not showing the other heroes. But it comes with a lot of code.

(to that I wanted to refer with my previous comment, but somehow didn't mentioned that I talk about this)

Comment (konecnyjakub) [reply units/heroes/Brainless_Heroes.cfg]: Well, I would omit hide_help for unit types. I do not mind if the player knows something ahead.

Comment (sevu) [reply units/heroes/Brainless_Heroes.cfg]: If so, then this is finished. (tough not used by the scenarios)
Have a look.

Comment (konecnyjakub) [reply units/heroes/Brainless_Heroes.cfg]: But I did. Since then you only pushed 1 commit concerning Boreas' animations which (I believe) cannot solve his biggest problem.

**Review by konecnyjakub — DISMISSED** [review]
Summing up issues that need to be addressed: 
Soulless Hero cannot advance to Running Hero (except Merrin). Running Hero needs [advancefrom], Soulless Hero's variation advances_to. Or copy WML from base unit instead of basing on it.
There is redundant drake variation (Pech is a gryphon).
Also, we need to actually use the new unit types in scenarios. I can do that myself after merging this.
And we should move old unit types for Merrin and Neckfrill to 1_compatibility.cfg but again, you do not have to do it here.

**Comment by sevu** [comment]
Would be done.
As for the merge conflict, looks like it is best resolved during merging.
Can you implement them in the scenarios starting from 3?

Edit: set this to the 2.0 milestone. Something is off in 1.12, I can't trace it down.

**Comment by konecnyjakub** [comment]
Imo, clearing the variable and adding unrenamable=yes should be a separate PR.
2 commits from this are also in #18. Intentional?
For the merge conflict, you do not have your other PR in this branch.

> Can you implement them in the scenarios starting from 3?

Sure, no problem.

> Something is off in 1.12

What exactly?

**Comment by sevu** [comment]
>Imo, clearing the variable and adding unrenamable=yes should be a separate PR.

Doesn't need to be here, is here because I came across this while doing the other things.

> 2 commits from this are also in #18. Intentional?

They should still be in 1.12.
I wonder if git can deal with the merge because they have the same id.
Well, if not it's easy to drop them here.

> For the merge conflict, you do not have your other PR in this branch.

I'd suggest manual merging, since it's trivial.

> Sure, no problem.

thanks

> What exactly?

Boreas has a red reference for the race.
The other units don't show the correct variation name nor the small_profile.
That all works in 1.13, eventually there is a workaround, but I would prefer to postpone it instead to work around it.

**Comment by konecnyjakub** [comment]
> I wonder if git can deal with the merge because they have the same id.
Well, if not it's easy to drop them here.

Dunno really, never experienced this. I would remove them from here, to be safe.

> Boreas has a red reference for the race.
The other units don't show the correct variation name nor the small_profile.

Happens with the pushed commits, right? Have you tried any solution proposed here: https://github.com/konecnyjakub/Danse_Macabre/pull/12#pullrequestreview-38782835? If so, how it went?

> I'd suggest manual merging, since it's trivial.

Removing that commit would be even easier :D since it is the last one in the branch. You can do just this:
```bash
git reset --hard HEAD~1
```

**Comment by sevu** [comment]
Your proposed solutions are all included, most were actually already a while back fixed.
Only copying WML I wanted to avoid.
I remember that a while back I had problems when I didn't override the inherited the variation_id in the hero variations, and was yesterday totally happy that this is fine now. This would be a workaround for _some_ problems. It's strange that boreas issues in the help differ a bit. Well, it's generally strange.
Oh, and this problems did affect placing the heroes with [unit], but not with debug.
(e.g. :create Soulless Hero & :unit variation=wose )
I suggest have a look with 1.13.8. and leave it here like it was always - or go with an object for the two remaining heroes, if you want to remove their unit types.

I'll merge upstream master into this PR after the other one is merged, or the next days, whichever is later.
Edit: btw, do_not_list got an additional functionality in 1.13, it doesn't list the unit in the list which you get with <shift> <c> in debug mode.

**Comment by konecnyjakub** [comment]
> Your proposed solutions are all included, most were actually already a while back fixed.

I see now, I must have missed that changes. Well, Soulless Hero's variations do not have advances_to. Iirc, that fixed the issue with (not) advancing for everyone but Boreas.

> Only copying WML I wanted to avoid.

Yes, I agree on this but if it is the only way, I would do that.

> btw, do_not_list got an additional functionality in 1.13, it doesn't list the unit in the list which you get with in debug mode.

I think that I have seen that. It is rather unpleasant.

**Comment by sevu** [comment]
> Yes, I agree on this but if it is the only way, I would do that.

The other way is using the wesnoth version where it is fixed
Boreas doesn't have any issues there.

The advances_to in the beginning of the Soulless Hero file is actually dead code because it's being overridden by the [advance_from] in the Running Hero (which is, in terms of overriding, rather a property of the Soulless Hero than of the Running Hero)
The - or a (?) - reason why we need the [advancefrom] is that it cleans up the mess in the wiki, which we have because of inheriting the [advancefrom] from Running to Running Hero.

I think said advance_to line disables advancing of the Soulless Hero to Running [nothero], but leaves a mess in the wiki.

**Comment by konecnyjakub** [comment]
> The other way is using the wesnoth version where it is fixed

Which is 1.13.x, right? For 1.12, we can copy the WML/do something else.

> Boreas doesn't have any issues there.

Oh, really? I tested only on Wesnoth 1.13.6 (and 1.12.6) and he was buggy as hell there. I could not get 1.13.7 from repositories (I am on Ubuntu).

> The advances_to in the beginning of the Soulless Hero file is actually dead code because it's being overridden by the [advance_from] in the Running Hero ... The - or a (?) - reason why we need the [advancefrom] is that it cleans up the mess in the wiki, which we have because of inheriting the [advancefrom] from Running to Running Hero.

Yes, I know that much.

**Comment by sevu** [comment]
> Which is 1.13.x, right? For 1.12, we can copy the WML/do something else.

I don't know, at least it is 1.13.8. I didn't come across a bug report or something...

**Comment by konecnyjakub** [comment]
1.13.8 is in repositories ... since yesterday. I have already downloaded that and will try this PR there. But not today.

**Comment by sevu** [comment]
yeah, has time

If you are in the mood for compiling, this are the options which I use when building with scons:
https://gist.github.com/sevu/308f301ca6536a4a84e5d4b1e66da292
you can install the dependencies with 
'apt-get build-dep wesnoth-1.13'
and then just type scons in the wesnoth base directory.

But I just read your answer, I think there's no need for compiling anymore.

File: units/heroes/Soulless_Heroes.cfg
Comment (konecnyjakub) [inline units/heroes/Soulless_Heroes.cfg]: A typo here. Should be hide_help.

**Comment by konecnyjakub** [comment]
Hmm, GitHub does not complain about the first two commits but it is kinda pointless to have them here now (after merging the other PR).
About Wesnoth 1.12, shall we copy the WML for Soulless Hero (inside #ifver)? That should solve the remaining issue (game play issue, that is). Unless the thing that solved it for 1.13 is going to be back-ported. But that is unlikely, I guess.

**Comment by sevu** [comment]
Not sure, you can give it a try.

**Comment by sevu** [comment]
added a new file for 1.12 support.
only the last commit is new, the other got only new IDs while rebasing
The nice thing is, because it has the same ids and variation ids there's no problem with deprecating it.

**Comment by konecnyjakub** [comment]
Looks fine but I will take a better look in game tomorrow.
Other (a bit unrelated) stuff: after merging this PR, I am going to release 1.3 as soon as possible. It will be the last version for Wesnoth 1.12 (unless something serious emerges). Is there anything else you would like to have in it?
Also, I have (finally) decided to give you push access to this repository. Well, after merging this.

**Comment by sevu** [comment]
Good to hear.
For the release, what do you think about #20? 
I think there is an engine bug, but I couldn't narrow it down yet.
( https://github.com/wesnoth/wesnoth/issues/1724 maybe the frame tag, maybe the alpha key or sth. else. )

Originally I wanted to get rid of the old animation syntax for the next release, but that can wait.

**Comment by konecnyjakub** [comment]
> For the release, what do you think about #20?

I did not find time to investigate it yet but will look into it.

File: units/heroes/1.12.cfg
Comment (konecnyjakub) [inline units/heroes/1.12.cfg]: Should be variation_id. As is now, we cannot create the variation, though it is present in help.

Comment (sevu) [reply units/heroes/1.12.cfg]: typos, typos, typos ...

**Comment by konecnyjakub** [comment]
Once that is fixed, this will work just fine. But it would be really better to create a new PR for changes to non-hero zombies (or re open the old one). (I will let the changes to scenarios go in.)

**Comment by sevu** [comment]
Updated the typo. It was a bit troublesome with the parallel PRs, because I wanted the commits from the other branch too, to not test without the bugfixes. Ended up in some more training with git :)

**Comment by konecnyjakub** [comment]
Hmm, it is not here. (Did you try push instead of force push? Or did you not push yet?)

> Ended up in some more training with git :)

That is always a good thing :).

**Comment by sevu** [comment]
oh, here it is, force-pushed, but didn't add it before.

**Comment by konecnyjakub** [comment]
Ok, now I will look into die animations of regular zombies. It does not matter for now where it is, though I would prefer separate PR for it.

**Comment by konecnyjakub** [comment]
The die animations are fine for Brain Eater and Battalion, but for Running Corpse they work just for drake/gryphon variation (not even the base unit). Reports 1 (?) image not found.

**Comment by sevu** [comment]
To much copy paste programming, updated.
Do you prefer force pushes or separate commits?

For gryphon/drake -if it's ok like now #20 isn't much of an issue

**Comment by konecnyjakub** [comment]
Force pushes, they make better history imo.

**Comment by sevu** [comment]
That's true.
I heard that other people may run into problems then, but I wasn't on the other side of things yet.

**Comment by konecnyjakub** [comment]
Well, for PRs, that is. You should not force-push to master and non-PR branches. GitHub even offers option to disable force pushes to certain branches (and I have turned it on here for master).

**Comment by konecnyjakub** [comment]
Works now :). Apply the fix to the other PR, please. Also the part "enable portraits for generic units from 1.13.8 on only (avoid inheriting before)" might go there (probably as separate commit). I will then merge that. It is better to have these things in another PR imo. If I am not mistaken, no rebase should be required here after that.

**Comment by sevu** [comment]
You mean you want to squash merge this PR?

**Comment by konecnyjakub** [comment]
No, I would like you to move those commits to #19.

**Comment by sevu** [comment]
Don't worry about that. In the end, both PRs are getting merged into master, so it ends up the same.
Github offers three merge options - with rebase merge you won't get a merge commit, and since there hasn't been work on master either, the commit history in master will look exactly like the one in this PR.

**Comment by konecnyjakub** [comment]
Yes, I know but 1 pull request should address 1 thing and those commits somehow does not fit among the others.

**Comment by sevu** [comment]
Well.. can do that, but ... basically I'll split the commits to branches and you merge them together like they are now .. Though I still miss the advantage which this would give to us now.
It may eventually have helped to review, since you have the code more split up, but were already past that.

**Review by konecnyjakub — APPROVED** [review]
Sure, there is no practical advantage at this point. Probably I should have insisted on it sooner. So, merging as is.

**Pull request merged**
